oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/PersistenceTimer.java

```diff
@@ -63,98 +70,149 @@ public void start(ModuleManager moduleManager, CoreModuleConfig moduleConfig) {
+                        }
+                        List<PrepareRequest> innerPrepareRequests = new ArrayList<>(5000);
+                        worker.buildBatchRequests(innerPrepareRequests);
+                        synchronized (prepareRequests) {
+                            prepareRequests.addAll(innerPrepareRequests);
+                            if (prepareRequests.size() >= maxSyncoperationNum) {
+                                prepareRequests.notify();
+                            }
+                        }
+                        worker.endOfRound(System.currentTimeMillis() - lastTime);
+                    } finally {
+                        timer.finish();
+                        countDownLatch.countDown();
                     }
-
-                    worker.buildBatchRequests(prepareRequests);
-
-                    worker.endOfRound(System.currentTimeMillis() - lastTime);
                 });
+            });
+
+            Future<?> batchFuture = batchExecutorService.submit(() -> {
+                List<Future<?>> results = new ArrayList<>();
+                while (true) {
+                    synchronized (prepareRequests) {
+                        if (prepareDone && CollectionUtils.isEmpty(prepareRequests)) {
+                            break;
+                        }
+                    }
 
-                if (debug) {
-                    log.info("build batch persistence duration: {} ms", System.currentTimeMillis() - startTime);
-                }
-            } finally {
-                timer.finish();
-            }
+                    synchronized (prepareRequests) {
+                        while (this.prepareRequests.size() < maxSyncoperationNum && !prepareDone) {
+                            try {
+                                this.prepareRequests.wait(1000);
+                            } catch (InterruptedException e) {
+                            }
+                        }
 
-            HistogramMetrics.Timer executeLatencyTimer = executeLatency.createTimer();
-            try {
-                List<List<PrepareRequest>> partitions = Lists.partition(prepareRequests, maxSyncoperationNum);
-                CountDownLatch countDownLatch = new CountDownLatch(partitions.size());
-                for (final List<PrepareRequest> partition : partitions) {
-                    executorService.submit(() -> {
+                        if (CollectionUtils.isEmpty(prepareRequests)) {
+                            continue;
+                        }
+                    }
+
+                    List<PrepareRequest> partition = null;
+                    synchronized (prepareRequests) {
+                        List<PrepareRequest> prepareRequestList = this.prepareRequests.subList(0, Math.min(maxSyncoperationNum, this.prepareRequests.size()));
+                        partition = new ArrayList<>(prepareRequestList);
+                        prepareRequestList.clear();
```

Comment:
prepareRequestList is created in L171, and only a reference, why should be clear?

Reply:
PrepareRequestList clear will remove range from `this.prepareRequests`. So we just need to take between 0 and maxSyncoperationNum from `this.prepareRequests` until `this.prepareRequests` is empty. 


```
this.prepareRequests = [0,1,2,3,4,5,6,7,8,9,10];

this.prepareRequests.subList(0,5).clear();

// after this.prepareRequests = [6,7,8,9,10];

```

Reply:
> prepareRequestList is created in L171, and only a reference, why should be clear?

I think we don't even need `prepareRequestList`, just 

```suggestion
                        partition = this.prepareRequests.subList(0, Math.min(maxSyncoperationNum, this.prepareRequests.size()));
```

because `partition` is not meant to be modified in `batchDAO.synchronous`, no?

Reply:
My point is the same. `partition` is enough. `#clear` is only relating to GC, but in Java, we don't need to do that.

Reply:
If we did not clear the sublist cause always get the same range from `this.prepareRequests` 
https://github.com/apache/skywalking/pull/7153#discussion_r656046059

Reply:
> If we did not clear the sublist cause always get the same range from `this.prepareRequests`
> [#7153 (comment)](https://github.com/apache/skywalking/pull/7153#discussion_r656046059)

Then this makes sense to me. 👍🏻